decentraland/unity-renderer
# feat: UI optimizations

## What does this PR change?

- Removes the refresh polling that degrades the performance when there are a lot of UIShapes, in its place now there's a unique coroutine that checks if there's a dirty UIShape on its queue.
- Removed a constant Canvas Rebuild caused by a TextShape being updated every frame with no changes on its data.

Fixes #2711

## Performance Comparison

### Old ( ICE Poker - The Stronghold ) [No Avatars]
```
 * PERFORMANCE SCORE (0-100) -> 64
 * lowest frame time -> 38.9ms
 * average frame time -> 45.8ms
 * highest frame time -> 100.0ms
 * 1 percentile frame time -> 39.3ms
 * 50 percentile frame time -> 44.2ms
 * 99 percentile frame time -> 100.0ms
 * error percentage -> ±1.5%
 * total hiccups (>0.05ms frames) -> 44 (6.72% of frames were hiccups)
 * total hiccups time (seconds) -> 3.076782
 * total frames -> 655
 * total frames time (seconds) -> 30.00663
```
### New

```
 * PERFORMANCE SCORE (0-100) -> 69
 * lowest frame time -> 35.3ms
 * average frame time -> 40.6ms
 * highest frame time -> 100.0ms
 * 1 percentile frame time -> 36.2ms
 * 50 percentile frame time -> 40.0ms
 * 99 percentile frame time -> 62.8ms
 * error percentage -> ±1.0%
 * total hiccups (>0.05ms frames) -> 14 (1.89% of frames were hiccups)
 * total hiccups time (seconds) -> 0.9778976
 * total frames -> 740
 * total frames time (seconds) -> 30.02143
```

## How to test

- Go to a scene with lots of UI elements like ICE Poker (-100,126) and make sure that they are correctly aligned when they get updated.
- Go to the same scene and before the UI loads, run outside the building and wait for all the tables to load, then enter the scene, the UI should appear without a massive hiccup


## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
